pkg/util/objectpool/objectpool.go

```diff
@@ -32,53 +32,51 @@ type PoolObject interface {
 }
 
 type (
+	// MultiPool manage multi Pool.
+	MultiPool struct {
+		pools sync.Map
+		lock  sync.Mutex
+		spec  *Spec
+	}
 	// Pool manage the PoolObject
 	Pool struct {
-		initSize     int                        // initial size
-		maxSize      int                        // max size
-		size         int                        // current size
-		new          func() (PoolObject, error) // create a new object, it must return a health object or err
-		store        chan PoolObject            // store the object
-		cond         *sync.Cond                 // when conditions are met, it wakes all goroutines waiting on sync.Cond
-		checkWhenGet bool                       // whether to health check when get PoolObject
-		checkWhenPut bool                       // whether to health check when put PoolObject
+		spec  *Spec
+		size  int             // current size
+		store chan PoolObject // store the object
+		cond  *sync.Cond      // when conditions are met, it wakes all goroutines waiting on sync.Cond
 	}
 
 	// Spec Pool's spec
 	Spec struct {
-		InitSize     int                        // initial size
-		MaxSize      int                        // max size
-		New          func() (PoolObject, error) // create a new object
-		CheckWhenGet bool                       // whether to health check when get PoolObject
-		CheckWhenPut bool                       // whether to health check when put PoolObject
+		InitSize     int                                           // initial size
+		MaxSize      int                                           // max size
+		New          func(ctx context.Context) (PoolObject, error) // create a new object, it must return a health object or err
+		CheckWhenGet bool                                          // whether to health check when get PoolObject
+		CheckWhenPut bool                                          // whether to health check when put PoolObject
 	}
 )
 
 // New returns a new pool
-func New(initSize, maxSize int, new func() (PoolObject, error)) *Pool {
-	return NewWithSpec(Spec{
+func New(initSize, maxSize int, new func(ctx context.Context) (PoolObject, error), ctx context.Context) *Pool {
+	return NewWithSpec(&Spec{
 		InitSize:     initSize,
 		MaxSize:      maxSize,
 		New:          new,
 		CheckWhenGet: true,
 		CheckWhenPut: true,
-	})
+	}, ctx)
 }
 
 // NewWithSpec returns a new pool
-func NewWithSpec(spec Spec) *Pool {
+func NewWithSpec(spec *Spec, ctx context.Context) *Pool {
```

Comment:
why do we need the `ctx` here?
I think the pool should not care about it, otherwise, the code will become too specific and not reusable.
can we find another solution to avoid it?

Reply:
In some scenarios, func `new` will require some input parameters to implement the business logic, and these parameters cannot be determined when the pool is created, such as the target address of the connection. so I envisage using ctx to pass variables, on the other hand, if the input params are not required in some scenarios, then input nil is ok. . Of course, we can also write anonymous functions to refer to context variables (I tried), but I think the code will be clearer.

Reply:
I failed to find a better solution, we can accept this for now, but please add a comment to say something about this, to let other developer know this may be changed later.